pkg/auth/oauth2.go

```diff
@@ -0,0 +1,178 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package auth
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/streamnative/pulsarctl/pkg/auth/oauth2"
+	"github.com/streamnative/pulsarctl/pkg/auth/oauth2/cache"
+	"github.com/streamnative/pulsarctl/pkg/auth/oauth2/store"
+	xoauth2 "golang.org/x/oauth2"
+	clock2 "k8s.io/utils/clock"
+)
+
+const (
+	TypeClientCredential = "client_credentials"
+	TypeDeviceCode       = "device_code"
+)
+
+type OAuth2Provider struct {
+	isCommandLineTools bool
+	clock              clock2.RealClock
+	issuer             oauth2.Issuer
+	store              store.Store
+	source             cache.CachingTokenSource
+	T                  http.RoundTripper
+}
+
+func NewAuthenticationOauth2(issuer oauth2.Issuer, store store.Store, types, keyFile string) (*OAuth2Provider, error) {
+	p := &OAuth2Provider{
+		clock:  clock2.RealClock{},
+		issuer: issuer,
+		store:  store,
+	}
+	err := p.Init(types, keyFile)
+	return p, err
+}
+
+func NewAuthenticationOAuth2WithParams(
+	commandLineTool bool,
+	types,
+	issueEndpoint,
+	clientID,
+	audience,
+	keyFile string) (*OAuth2Provider, error) {
+
+	issuer := oauth2.Issuer{
+		IssuerEndpoint: issueEndpoint,
+		ClientID:       clientID,
+		Audience:       audience,
+	}
+
+	var st store.Store
+	if commandLineTool {
+		keyringStore, err := store.MakeKeyringStore()
+		if err != nil {
+			return nil, err
+		}
+		st = keyringStore
+	} else {
+		st = store.NewMemoryStore()
+	}
+
+	return NewAuthenticationOauth2(issuer, st, types, keyFile)
+}
+
+func (o *OAuth2Provider) Load() {
+
+}
+
+func (o *OAuth2Provider) Init(types, keyFile string) error {
+	var grant *oauth2.AuthorizationGrant
+	switch types {
+	case TypeClientCredential:
+		flow, err := oauth2.NewDefaultClientCredentialsFlow(o.issuer, keyFile)
+		if err != nil {
+			return err
+		}
+		if o.isCommandLineTools {
+			grant, err = o.store.LoadGrant(o.issuer.Audience)
+			if err != nil {
+				return err
+			}
+		} else {
+			grant, err := flow.Authorize()
+			if err != nil {
+				return err
+			}
+			err = o.store.SaveGrant(o.issuer.Audience, *grant)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
I think it would be better to not vary the behavior here with `isCommandLineTools`.   The uniform thing to do here is `LoadGrant` with the assumption that the store has already been prepared.

To use pulsarctl as a library, the application could call `NewAuthenticationOauth2(issuer, store)` and should handle filling the store with grants.  

To use pulsarctl as a standalone tool, it should call `NewAuthenticationOAuth2WithParams` based on the config file.  As seen here.
https://github.com/apache/pulsar-client-go/pull/307/files